src/Features/Core/Portable/CodeAnalysisSuggestions/CodeAnalysisSuggestionsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,183 @@
+        if (codeFixService == null)
+            return;
+
+        var (actions, totalViolations) = await GetCodeAnalysisSuggestionActionsAsync(ruleConfigData, document, codeFixService, cancellationToken).ConfigureAwait(false);
+        if (actions.Length > 0)
+        {
+            Debug.Assert(totalViolations > 0);
+
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    string.Format(FeaturesResources.Code_analysis_improvements_0, totalViolations),
+                    actions,
+                    isInlinable: false,
+                    CodeActionPriority.Low),
+                span);
+        }
+    }
+
+    private static async Task<(ImmutableArray<CodeAction> Actions, int TotalViolations)> GetCodeAnalysisSuggestionActionsAsync(
+        ImmutableArray<(string, ImmutableDictionary<string, ImmutableArray<DiagnosticData>>)> configData,
+        Document document,
+        ICodeFixService codeFixService,
+        CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var location = root.GetLocation();
+
+        using var _1 = ArrayBuilder<CodeAction>.GetInstance(out var actionsBuilder);
+        using var _2 = ArrayBuilder<CodeAction>.GetInstance(out var nestedActionsBuilder);
+        using var _3 = ArrayBuilder<CodeAction>.GetInstance(out var nestedNestedActionsBuilder);
+        var totalViolations = 0;
+        foreach (var (category, diagnosticsById) in configData)
+        {
+            var totalViolationsForCategory = 0;
+            foreach (var (id, diagnostics) in diagnosticsById)
+            {
+                Debug.Assert(diagnostics.All(d => string.Equals(d.Category, category, StringComparison.OrdinalIgnoreCase)));
+
+                var (diagnosticData, documentForFix) = GetPreferredDiagnosticAndDocument(diagnostics, document);
+                var diagnostic = await diagnosticData.ToDiagnosticAsync(document.Project, cancellationToken).ConfigureAwait(false);
+                if (SuppressionHelpers.IsNotConfigurableDiagnostic(diagnostic))
+                    continue;
+
+                if (documentForFix != null)
+                {
+                    var codeFixCollection = await codeFixService.GetDocumentFixAllForIdInSpanAsync(documentForFix, diagnostic.Location.SourceSpan, id, diagnostic.Severity, CodeActionOptions.DefaultProvider, cancellationToken).ConfigureAwait(false);
+                    if (codeFixCollection != null)
+                    {
+                        nestedNestedActionsBuilder.AddRange(codeFixCollection.Fixes.Select(f => f.Action));
+                    }
+                    else
+                    {
+                        // This diagnostic has no code fix, or fix application failed.
+                        // TODO: Can we add some code action such that it shows a preview of the diagnostic span with squiggle?
```

Comment:
intentional that we're not returning?

Reply:
Yes, we would still like to show `Configure severity` action for such diagnostics